services/galley/src/Galley/Data.hs

```diff
@@ -691,17 +712,26 @@ privateRole = PrivateAccessRole
 privateOnly :: Set Access
 privateOnly = Set [PrivateAccess]
 
-toMappedOrLocalId :: (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain) -> MappedOrLocalId a
+type MappedOrLocalIdRow a = (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain)
+
+toMappedOrLocalId :: (Log.MonadLogger m, MonadThrow m) => MappedOrLocalIdRow a -> m (MappedOrLocalId a)
 toMappedOrLocalId = \case
-  (Id mappedId, Just remoteId, Just domain) ->
-    Mapped (IdMapping (Id mappedId) (Qualified remoteId domain))
-  (Id localId, Nothing, Nothing) ->
-    Local (Id localId)
-  (Id localId, _, _) ->
-    -- TODO(mheinzel): return either, raise error
-    Local (Id localId)
-
-fromMappedOrLocalId :: MappedOrLocalId a -> (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain)
+  (mappedId, Just remoteId, Just domain) ->
+    pure $ Mapped (IdMapping (coerce mappedId) (Qualified remoteId domain))
+  (localId, Nothing, Nothing) ->
+    pure $ Local (coerce localId)
+  invalid -> do
+    -- This should never happen as we always write rows with either both or none of these
+    -- values.
+    -- FUTUREWORK: we could try to recover from this situation by checking if an ID mapping
+    -- for this mapped ID exists (and potentially even repair the row). At the moment, the
+    -- problem seems unlikely enough not to warrant the complexity, though.
+    -- In some cases it could also be better not to fail, but skip this entry, e.g. when
+    -- deleting a user, we should remove him from all conversations we can, not stop halfway.
```

Comment:
No, it's helpful.

I first tried returning the `Either` but it was much more complicated, because a lot of functions in this module now had to deal with both the `m` and the `Either` and they were still general enough that they couldn't make the decision how to deal with it, so it needed bubbling up another level into the `API` modules where a lot of functions were affected.

It would have been the most precise approach, but also most cumbersome. This one is kind of in the middle. Still undecided between all three of them (although the `Either` would be most work to do, quite a lot of plumbing would need to be written).

I'll wait for a third opinion from @arianvp if possible.

Reply:
After some more discussion, I left it as it is now.